include/deal.II/base/floating_point_comparator.h

```diff
@@ -118,82 +147,143 @@ FloatingPointComparator<VectorizedArrayType>::operator()(
     return false;
   else
     for (unsigned int i = 0; i < s1; ++i)
-      if (v1[i] < v2[i] - tolerance)
+      if (this->operator()(v1[i], v2[i]))
         return true;
-      else if (v1[i] > v2[i] + tolerance)
+      else if (this->operator()(v2[i], v1[i]))
         return false;
   return false;
 }
 
 
 
-template <typename VectorizedArrayType>
+template <typename Number>
 bool
-FloatingPointComparator<VectorizedArrayType>::operator()(
-  const Tensor<1, width, Number> &t1,
-  const Tensor<1, width, Number> &t2) const
+FloatingPointComparator<Number>::operator()(
+  const Tensor<1, width, ScalarNumber> &t1,
+  const Tensor<1, width, ScalarNumber> &t2) const
 {
   for (unsigned int k = 0; k < width; ++k)
-    if (t1[k] < t2[k] - tolerance)
+    if (this->operator()(t1[k], t2[k]))
       return true;
-    else if (t1[k] > t2[k] + tolerance)
+    else if (this->operator()(t2[k], t1[k]))
       return false;
   return false;
 }
 
 
 
-template <typename VectorizedArrayType>
+template <typename Number>
 template <int dim>
 bool
-FloatingPointComparator<VectorizedArrayType>::operator()(
-  const Tensor<1, dim, Tensor<1, width, Number>> &t1,
-  const Tensor<1, dim, Tensor<1, width, Number>> &t2) const
+FloatingPointComparator<Number>::operator()(
+  const Tensor<1, dim, Tensor<1, width, ScalarNumber>> &t1,
+  const Tensor<1, dim, Tensor<1, width, ScalarNumber>> &t2) const
 {
   for (unsigned int d = 0; d < dim; ++d)
     for (unsigned int k = 0; k < width; ++k)
-      if (t1[d][k] < t2[d][k] - tolerance)
+      if (this->operator()(t1[d][k], t2[d][k]))
         return true;
-      else if (t1[d][k] > t2[d][k] + tolerance)
+      else if (this->operator()(t2[d][k], t1[d][k]))
         return false;
```

Comment:
Now that I see this implementation, I wonder if we should make the function 
```c++
FloatingPointComparator<Number>::operator()(
  const Tensor<1, width, ScalarNumber> &t1,
  const Tensor<1, width, ScalarNumber> &t2) const
```
a template, i.e.,
```c++
template <typename OtherNumber>
FloatingPointComparator<Number>::operator()(
  const Tensor<1, width, OtherNumber> &t1,
  const Tensor<1, width, OtherNumber> &t2) const
```
when then channels everything through the other function? That would allow us to reduce the number of function declarations. Maybe, we could add a second template argument for the `rank` of the tensor and even get the other function through this code? This was my code initialize, so I take the blame, but I think I see some opportunity for simplification.

Reply:
@kronbichler Last commit. What do you think?

Reply:
> [...] channels everything through the other function? That would allow us to reduce the number of function declarations.

I feel embarrassed by my suggestion. The performance tests suggest that for the step-37 benchmark, the time for `setup_matrix_free` went from 0.49 seconds to 0.52 seconds (statistically secured). The reason is that we keep checking both sides of the comparison (necessarily) for each nested level, so for something like `std::array<Tensor<2,dim>,4>` we keep doing around 10 times as much work (I did not count exactly to figure out if the factor is 6, 8, 9, 18 or something like that). Unfortunately, I do not see compilers throw away all duplicate comparisons for us, so we will need to roll back and introduce the specialized code.

Incidentally, what I suggested here is similar to using recursive function calls for computing the Fibonacci numbers, i.e. compute fib(n) by calling separately the functions fib(n-1) and fib(n-2). This is something I consider utterly stupid, because it  makes the complexity O(2^n) rather than O(n), and no sane person should do this.

Reply:
I will do a PR with a fix in a few hours.